File: Flink/VoipStream/src/main/java/VoipStream/ACD.java
/**************************************************************************************
 *  Copyright (c) 2019- Gabriele Mencagli and Andrea Cardaci
 *  
 *  This file is part of StreamBenchmarks.
 *  
 *  StreamBenchmarks is free software dual licensed under the GNU LGPL or MIT License.
 *  You can redistribute it and/or modify it under the terms of the
 *    * GNU Lesser General Public License as published by
 *      the Free Software Foundation, either version 3 of the License, or
 *      (at your option) any later version
 *    OR
 *    * MIT License: https://github.com/ParaGroup/StreamBenchmarks/blob/master/LICENSE.MIT
 *  
 *  StreamBenchmarks is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *  You should have received a copy of the GNU Lesser General Public License and
 *  the MIT License along with WindFlow. If not, see <http://www.gnu.org/licenses/>
 *  and <http://opensource.org/licenses/MIT/>.
 **************************************************************************************
 */

package VoipStream;

import org.apache.flink.api.common.functions.RichFlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple6;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.util.Collector;
import org.slf4j.Logger;
import common.CallDetailRecord;
import common.Constants;
import common.ScorerMap;
import util.Log;

import java.util.Map;

public class ACD extends RichFlatMapFunction<
        Tuple6<Integer, Long, String, Long, Double, CallDetailRecord>,
        Tuple6<Integer, Long, String, Long, Double, CallDetailRecord>> {
    private static final Logger LOG = Log.get(ACD.class);

    private double avg;
    private ScorerMap scorerMap;

    @Override
    public void open(Configuration parameters) {

        scorerMap = new ScorerMap(new int[]{ScorerMap.CT24, ScorerMap.ECR24});
    }

    @Override
    public void flatMap(Tuple6<Integer, Long, String, Long, Double, CallDetailRecord> tuple, Collector<Tuple6<Integer, Long, String, Long, Double, CallDetailRecord>> collector) throws Exception {
        LOG.debug("tuple in: {}", tuple);

        long timestamp = tuple.f1;
        int source = tuple.f0;

        if (source == ScorerMap.GlobalACD) {
            avg = tuple.f4;
        } else {
            CallDetailRecord cdr = tuple.f5;
            String number = tuple.f2;
            long answerTimestamp = tuple.f3;
            double rate = tuple.f4;

            String key = String.format("%s:%d", number, answerTimestamp);

            Map<String, ScorerMap.Entry> map = scorerMap.getMap();
            if (map.containsKey(key)) {
                ScorerMap.Entry entry = map.get(key);
                entry.set(source, rate);

                if (entry.isFull()) {
                    // calculate the score for the ratio
                    double ratio = (entry.get(ScorerMap.CT24) / entry.get(ScorerMap.ECR24)) / avg;
                    double score = ScorerMap.score(Constants.ACD_THRESHOLD_MIN, Constants.ACD_THRESHOLD_MAX, ratio);

                    Tuple6<Integer, Long, String, Long, Double, CallDetailRecord> out = new Tuple6<>(ScorerMap.ACD, timestamp, number, answerTimestamp, score, cdr);
                    collector.collect(out);
                    LOG.debug("tuple out: {}", out);

                    map.remove(key);
                }
            } else {
                ScorerMap.Entry entry = scorerMap.newEntry();
                entry.set(source, rate);
                map.put(key, entry);
            }
        }
    }
}

File: WindFlow/VoipStream/lib/voip_stream/acd.cpp
/**************************************************************************************
 *  Copyright (c) 2019- Gabriele Mencagli and Andrea Cardaci
 *  
 *  This file is part of StreamBenchmarks.
 *  
 *  StreamBenchmarks is free software dual licensed under the GNU LGPL or MIT License.
 *  You can redistribute it and/or modify it under the terms of the
 *    * GNU Lesser General Public License as published by
 *      the Free Software Foundation, either version 3 of the License, or
 *      (at your option) any later version
 *    OR
 *    * MIT License: https://github.com/ParaGroup/StreamBenchmarks/blob/master/LICENSE.MIT
 *  
 *  StreamBenchmarks is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *  You should have received a copy of the GNU Lesser General Public License and
 *  the MIT License along with WindFlow. If not, see <http://www.gnu.org/licenses/>
 *  and <http://opensource.org/licenses/MIT/>.
 **************************************************************************************
 */

#include "acd.hpp"
#include "constants.hpp"
#include "util/log.hpp"

namespace voip_stream {

ACD::ACD()
    : avg_(0)
    , scorer_map_({SOURCE_CT24, SOURCE_ECR24})
{}

void ACD::operator ()(const FilterTuple &tuple, wf::Shipper<FilterTuple> &shipper, wf::RuntimeContext &rc)
{
    DEBUG_LOG("acd::tuple " << tuple);

    if (tuple.source == SOURCE_GlobalACD) {
        avg_ = tuple.rate;
    } else {
        auto key = tuple.cdr.calling_number + ':' + std::to_string(tuple.cdr.answer_timestamp);
        auto &map = scorer_map_.get_map();
        auto it = map.find(key);

        if (it != map.end()) {
            auto &entry = it->second;
            entry.set(tuple.source, tuple.rate);

            if (entry.is_full()) {
                // calculate the score for the ratio
                double ratio = (entry.get(SOURCE_CT24) / entry.get(SOURCE_ECR24)) / avg_;
                double score = ScorerMap::score(ACD_THRESHOLD_MIN, ACD_THRESHOLD_MAX, ratio);

                FilterTuple result;
                result.ts = tuple.ts;
                result.source = SOURCE_ACD;
                result.cdr = tuple.cdr;
                result.rate = score;
                shipper.push(std::move(result));

                map.erase(key);
            }
        } else {
            ScorerMap::Entry entry = scorer_map_.new_entry();
            entry.set(tuple.source, tuple.rate);
            map.insert(std::make_pair(key, entry));
        }
    }
}

}
